compiler/generator_php.c

```diff
@@ -0,0 +1,1406 @@
+    }
+
+    wsetlab_begin(g, g->failure_label);
+    generate(g, p->left);
+    wsetlab_end(g,g->failure_label);
+    g->unreachable = false;
+
+    if (savevar) {
+        str_delete(savevar);
+    }
+}
+
+static void generate_set(struct generator * g, struct node * p) {
+    write_comment(g, p);
+    writef(g, "~M$this->~V = true;~N", p);
+}
+
+static void generate_unset(struct generator * g, struct node * p) {
+    write_comment(g, p);
+    writef(g, "~M$this->~V = false;~N", p);
+}
+
+static void generate_fail(struct generator * g, struct node * p) {
+    write_comment(g, p);
+    generate(g, p->left);
+    if (!g->unreachable) write_failure(g);
+}
+
+/* generate_test() also implements 'reverse' */
+
+static void generate_test(struct generator * g, struct node * p) {
+    struct str * savevar = NULL;
+    if (K_needed(g, p->left)) {
+        savevar = vars_newname(g);
+    }
+
+    write_comment(g, p);
+
+    if (savevar) {
+        write_savecursor(g, p, savevar);
+    }
+
+    generate(g, p->left);
+
+    if (savevar) {
+        if (!g->unreachable) {
+            write_restorecursor(g, p, savevar);
+        }
+        str_delete(savevar);
+    }
+}
+
+static void generate_do(struct generator * g, struct node * p) {
+    struct str * savevar = NULL;
+    if (K_needed(g, p->left)) {
+        savevar = vars_newname(g);
+    }
+
+    write_comment(g, p);
+    if (savevar) write_savecursor(g, p, savevar);
+
+    if (p->left->type == c_call) {
+        /* Optimise do <call> */
+        write_comment(g, p->left);
+        writef(g, "~M$this->~V();~N",p->left);
+    } else {
+        g->failure_label = new_label(g);
+        str_clear(g->failure_str);
+
+        wsetlab_begin(g, g->failure_label);
+        generate(g, p->left);
+        wsetlab_end(g,g->failure_label);
+        g->unreachable = false;
+    }
+
+    if (savevar) {
+        write_restorecursor(g, p, savevar);
+        str_delete(savevar);
+    }
+}
+
+static void generate_next(struct generator * g, struct node * p) {
+    write_comment(g, p);
+    write_check_limit(g, p);
+    write_inc_cursor(g, p);
+}
+
+static void generate_GO_grouping(struct generator * g, struct node * p, int is_goto, int complement) {
+    write_comment(g, p);
+
+    struct grouping * q = p->name->grouping;
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->S[1] = complement ? "in" : "out";
+    g->I[0] = q->smallest_ch;
+    g->I[1] = q->largest_ch;
+    write_failure_if(g, "!$this->go_~S1_grouping~S0($this->~V, ~I0, ~I1)", p);
+    if (!is_goto) {
+        if (p->mode == m_forward)
+            w(g, "~M$this->cursor++;~N");
+        else
+            w(g, "~M$this->cursor--;~N");
+    }
+}
+
+static void generate_GO(struct generator * g, struct node * p, int style) {
+    write_comment(g, p);
+
+    int a0 = g->failure_label;
+    struct str * a1 = str_copy(g->failure_str);
+
+    int end_unreachable = false;
+    int golab = new_label(g);
+    g->I[0] = golab;
+    write_comment(g, p);
```

Comment:
We already wrote the comment above so this looks like it'll give a duplicate comment (needs `-comments` to enable comments).

Reply:
Removed in 4d53f9b